pyvista/core/_validation/check.py

```diff
@@ -546,75 +652,215 @@ def check_shape(
         shape = [shape]
 
-    array_shape = arr.shape
-    for shp in shape:
-        shp = _validate_shape_value(shp)
-        if _shape_is_allowed(array_shape, shp):
+    array_shape = _ArrayLikeWrapper(array).shape
+    for input_shape in shape:
+        input_shape = _validate_shape_value(input_shape)
+        if _shape_is_allowed(array_shape, input_shape):
             return
 
-    msg = f"{name} has shape {arr.shape} which is not allowed. "
+    msg = f"{name} has shape {array_shape} which is not allowed. "
     if len(shape) == 1:
         msg += f"Shape must be {shape[0]}."
     else:
         msg += f"Shape must be one of {shape}."
     raise ValueError(msg)
 
 
-def check_number(num, /, *, name='Object'):
-    """Check if an object is an instance of ``Number``.
+def check_ndim(
+    array: _ArrayLikeOrScalar[_NumberType],
+    /,
+    ndim: Union[int, Sequence[int]],
+    *,
+    name: str = "Array",
+):
+    """Check if an array has the specified number of dimensions.
 
-    A number is any instance of ``numbers.Number``, e.g.  ``int``,
-    ``float``, and ``complex``.
+    A value of ``0`` is returned for scalar values.
+
+    Parameters
+    ----------
+    array : float | Array[float]
+        Number or array to check.
+
+    ndim : int | Sequence[int], optional
+        A single dimension or a sequence of allowable dimensions. If an
+        integer, the array must have this number of dimension(s). If a
+        sequence, the array must have at least one of the specified number
+        of dimensions.
+
+    name : str, default: "Array"
+        Variable name to use in the error messages if any are raised.
+
+    Raises
+    ------
+    ValueError
+        If the array does not have the required number of dimensions.
+
+    See Also
+    --------
+    check_length
+    check_shape
+
+    Examples
+    --------
+    Check if an array is one-dimensional
+
+    >>> import numpy as np
+    >>> from pyvista import _validation
+    >>> _validation.check_ndim([1, 2, 3], ndim=1)
+
+    Check if an array is two-dimensional or a scalar.
+
+    >>> _validation.check_ndim(1, ndim=(0, 2))
+
+    """
+    if not isinstance(ndim, Sequence):
+        ndim_: Sequence[int] = [ndim]
+    else:
+        ndim_ = ndim
+
+    array_ndim = _ArrayLikeWrapper(array).ndim
+    if array_ndim not in ndim_:
+        check_ndim(ndim, [0, 1], name='ndim')
+
+        if len(ndim_) == 1:
+            check_integer(ndim_[0], strict=True, name='ndim')
+            expected = f"{ndim}"
+        else:
+            check_integer(ndim, strict=True, name='ndim')
+            expected = f"one of {ndim}"
+        msg = (
+            f"{name} has the incorrect number of dimensions. "
+            f"Got {array_ndim}, expected {expected}."
+        )
+        raise ValueError(msg)
+
+
+def check_number(
+    num: Union[float, int, complex, np.number, Number],  # type: ignore[type-arg]
+    /,
+    *,
+    definition: Literal['abstract', 'builtin', 'numpy'] = 'abstract',
```

Comment:
At one point it wasn't clear to me if support for complex numbers was needed. This function was generalized accordingly. But, the must_be_real check is a bit redundant... just use `check_real` if you need to, no? Having `must_be_real` here with default True (same as validate_array) is to make it explicit to the user though. Not sure if it should be changed.

Also, I went a bit mad because of the differences in definitions of what a number is in python. Hence the `definition` kwarg. But, it kind of complicates things unnecesarily. 

I think maybe aligning the definition of `number` with the `_NumberType` typevar is perhaps the way to go, i.e. python float or int, or numpy float or int. We could directly use `_is_Number` from `_type_guards.py` if that's the case and everything will be self-consistent.

Reply:
It's probably fine as is, can be changed later if needed.